File: ts/project/.sdk/tm/scala/feature/TestFeature.scala
package SCALAPACKAGE.feature

import java.util.{LinkedHashMap, List => JList, Map => JMap}
import java.util.concurrent.ThreadLocalRandom
import java.util.function.Supplier
import SCALAPACKAGE.core.{Context, FetcherFn, Helpers, Operation, SdkClient}
import SCALAPACKAGE.utility.struct.Struct

// In-memory mock transport for offline testing. Serves entity fixtures
// (options.entity) through the normal pipeline, and optionally simulates
// network conditions via the `net` block.
class TestFeature extends BaseFeature("test", "0.0.1", true) {

  private var client: SdkClient = null
  private var options: JMap[String, Object] = null
  private var netcalls: Int = 0

  override def init(ctx: Context, options: JMap[String, Object]): Unit = {
    this.client = ctx.client
    this.options = options

    val entity = Helpers.toMapAny(Struct.getprop(options, "entity"))

    this.client.mode = "test"

    // Ensure entity ids are correct.
    Struct.walk(entity, (key: String, valx: Object, parent: Object, path: java.util.List[String]) => {
      if (path != null && path.size() == 2 && valx.isInstanceOf[JMap[_, _]] && key != null) {
        valx.asInstanceOf[JMap[String, Object]].put("id", key)
      }
      valx
    })

    val entities: JMap[String, Object] = if (entity == null) new LinkedHashMap[String, Object]() else entity

    val testFetcher: FetcherFn = (fctx, fullurl, fetchdef) => serve(fctx, entities)

    val net = Helpers.toMapAny(Struct.getprop(options, "net"))
    if (net == null) ctx.utility.fetcher = testFetcher
    else ctx.utility.fetcher = makeNetsim(net, testFetcher)
  }

  private def respond(status: Int, data: Object, extra: JMap[String, Object]): JMap[String, Object] = {
    val js: Supplier[Object] = () => data
    val out = new LinkedHashMap[String, Object]()
    out.put("status", java.lang.Integer.valueOf(status))
    out.put("statusText", "OK")
    out.put("json", js)
    out.put("body", "not-used")
    if (extra != null) out.putAll(extra)
    out
  }

  private def extra(key: String, v: Object): JMap[String, Object] = {
    val out = new LinkedHashMap[String, Object]()
    out.put(key, v)
    out
  }

  private def resolveMatch(ctx: Context, explicit: JMap[String, Object]): JMap[String, Object] = {
    if (explicit != null && !explicit.isEmpty) return explicit
    val srcs = Array(ctx.matchData, ctx.data)
    var i = 0
    while (i < srcs.length) {
      val src = srcs(i)
      if (src != null) {
        val v = Struct.getprop(src, "id", null)
        if (v != null && !("__UNDEFINED__" == v)) {
          val out = new LinkedHashMap[String, Object]()
          out.put("id", v)
          return out
        }
      }
      i += 1
    }
    new LinkedHashMap[String, Object]()
  }

  private def serve(ctx: Context, entity: JMap[String, Object]): Object = {
    val op = ctx.op
    var entmap = Helpers.toMapAny(Struct.getprop(entity, op.entity))
    if (entmap == null) entmap = new LinkedHashMap[String, Object]()

    if ("load" == op.name) {
      val args = buildArgs(ctx, op, resolveMatch(ctx, ctx.reqmatch))
      val found = Struct.select(entmap, args)
      val ent = Struct.getelem(found, java.lang.Integer.valueOf(0))
      if (ent == null) return respond(404, null, extra("statusText", "Not found"))
      Struct.delprop(ent, "$KEY")
      val out = Struct.clone(ent)
      respond(200, out, null)
    } else if ("list" == op.name) {
      val args = buildArgs(ctx, op, ctx.reqmatch)
      val found = Struct.select(entmap, args)
      if (found == null) return respond(404, null, extra("statusText", "Not found"))
      val it = found.iterator()
      while (it.hasNext) Struct.delprop(it.next(), "$KEY")
      val out = Struct.clone(found)
      respond(200, out, null)
    } else if ("update" == op.name) {
      var updateMatch = new LinkedHashMap[String, Object]()
      if (ctx.reqdata != null) {
        if (ctx.reqdata.containsKey("id")) updateMatch.put("id", ctx.reqdata.get("id"))
        if (op.alias != null) {
          Struct.getprop(op.alias, "id") match {
            case aliasId: String if ctx.reqdata.containsKey(aliasId) => updateMatch.put(aliasId, ctx.reqdata.get(aliasId))
            case _ =>
          }
        }
      }
      if (updateMatch.isEmpty) updateMatch = resolveMatch(ctx, new LinkedHashMap[String, Object]()).asInstanceOf[LinkedHashMap[String, Object]]
      val args = buildArgs(ctx, op, updateMatch)
      val found = Struct.select(entmap, args)
      var ent = Struct.getelem(found, java.lang.Integer.valueOf(0))
      if (ent == null && entmap != null) {
        val vit = entmap.values().iterator()
        var brk = false
        while (vit.hasNext && !brk) {
          vit.next() match { case e: JMap[_, _] => ent = e; brk = true; case _ => }
        }
      }
      if (ent == null) return respond(404, null, extra("statusText", "Not found"))
      ent match { case m: JMap[_, _] if ctx.reqdata != null => m.asInstanceOf[JMap[String, Object]].putAll(ctx.reqdata); case _ => }
      Struct.delprop(ent, "$KEY")
      val out = Struct.clone(ent)
      respond(200, out, null)
    } else if ("remove" == op.name) {
      val args = buildArgs(ctx, op, resolveMatch(ctx, ctx.reqmatch))
      val found = Struct.select(entmap, args)
      val ent = Struct.getelem(found, java.lang.Integer.valueOf(0))
      ent match {
        case m: JMap[_, _] =>
          val id = Struct.getprop(m, "id", null)
          Struct.delprop(entmap, id)
        case _ =>
      }
      respond(200, null, null)
    } else if ("create" == op.name) {
      buildArgs(ctx, op, ctx.reqdata)
      var id = ctx.utility.param(ctx, "id")
      if (id == null) {
        val r = ThreadLocalRandom.current()
        id = String.format("%04x%04x%04x%04x",
          java.lang.Integer.valueOf(r.nextInt(0x10000)), java.lang.Integer.valueOf(r.nextInt(0x10000)),
          java.lang.Integer.valueOf(r.nextInt(0x10000)), java.lang.Integer.valueOf(r.nextInt(0x10000)))
      }

      val ent = Struct.clone(ctx.reqdata)
      ent match {
        case entm0: JMap[_, _] =>
          val entm = entm0.asInstanceOf[JMap[String, Object]]
          entm.put("id", id)
          id match { case s: String => entmap.put(s, entm); case _ => }
          Struct.delprop(entm, "$KEY")
          val out = Struct.clone(entm)
          respond(200, out, null)
        case _ => respond(200, ent, null)
      }
    } else {
      respond(404, null, extra("statusText", "Unknown operation"))
    }
  }

  private def makeNetsim(net: JMap[String, Object], inner: FetcherFn): FetcherFn = {
    this.netcalls = 0
    (ctx, url, fetchdef) => {
      this.netcalls += 1
      val call = this.netcalls

      if (FeatureOptions.foptBool(net, "offline", false)) {
        netSleep(net, pickLatency(net))
        throw ctx.makeError("netsim_offline", "Simulated network offline (URL was: \"" + url + "\")")
      }
      if (call <= FeatureOptions.foptInt(net, "errorTimes", 0)) {
        netSleep(net, pickLatency(net))
        throw ctx.makeError("netsim_conn", "Simulated connection error (call " + call + ")")
      }
      if (call <= FeatureOptions.foptInt(net, "failTimes", 0)) {
        netSleep(net, pickLatency(net))
        val status = FeatureOptions.foptInt(net, "failStatus", 503)
        val js: Supplier[Object] = () => null
        val out = new LinkedHashMap[String, Object]()
        out.put("status", java.lang.Integer.valueOf(status))
        out.put("statusText", "Simulated Failure")
        out.put("body", "not-used")
        out.put("json", js)
        out.put("headers", new LinkedHashMap[String, Object]())
        out
      } else {
        netSleep(net, pickLatency(net))
        inner(ctx, url, fetchdef)
      }
    }
  }

  private def pickLatency(net: JMap[String, Object]): Int = {
    net.get("latency") match {
      case null => 0
      case lm0: JMap[_, _] =>
        val lm = lm0.asInstanceOf[JMap[String, Object]]
        val min = FeatureOptions.foptInt(lm, "min", 0)
        val max = FeatureOptions.foptInt(lm, "max", min)
        if (max <= min) min else min + ((max - min) >> 1)
      case _ =>
        val fixed = FeatureOptions.foptInt(net, "latency", 0)
        if (fixed < 0) 0 else fixed
    }
  }

  private def netSleep(net: JMap[String, Object], ms: Int): Unit = {
    if (ms <= 0) return
    FeatureOptions.foptSleep(net).accept(ms)
  }

  private def buildArgs(ctx: Context, op: Operation, args: JMap[String, Object]): Object = {
    val opname = op.name

    val points = Struct.getpath(ctx.config,
      java.util.List.of("entity", if (ctx.entity == null) "" else ctx.entity.getName(), "op", opname, "points"))
    val point = Struct.getelem(points, java.lang.Integer.valueOf(-1))

    val paramsPath = Struct.getpath(point, java.util.List.of("args", "params"))
    val reqdParams = Struct.select(paramsPath, Struct.jm("reqd", java.lang.Boolean.TRUE))
    val reqd = Struct.transform(reqdParams, Struct.jt("`$EACH`", "", "`$KEY.name`"))

    val qand = Struct.jt()
    val q = Struct.jm("`$AND`", qand)

    if (args != null) {
      val kit = Struct.keysof(args).iterator()
      while (kit.hasNext) {
        val key = kit.next()
        val isId = "id" == key
        val selected = Struct.select(reqd, key)
        val isReqd = !Struct.isempty(selected)

        if (isId || isReqd) {
          val v = ctx.utility.param(ctx, key)
          val ka = Struct.getprop(op.alias, key)

          val qor = Struct.jt(Struct.jm(key, v))
          ka match { case s: String => qor.add(Struct.jm(s, v)); case _ => }

          qand.add(Struct.jm("`$OR`", qor))
        }
      }
    }

    if (ctx.ctrl.explain != null) ctx.ctrl.explain.put("test", Struct.jm("query", q))

    q
  }
}

File: ts/project/.sdk/tm/csharp/feature/TestFeature.cs
// In-memory mock transport for testing without a live server. Serves the
// per-entity fixture data supplied via the test feature options (`entity`)
// and answers CRUD requests the way a REST server would. An optional `net`
// block layers deterministic network-behaviour simulation (latency,
// first-N failures, connection errors, offline) over the mock.

using Voxgig.Struct;

using static ProjectNameSdk.Feature.FeatureOptions;

namespace ProjectNameSdk.Feature;

public class TestFeature : BaseFeature
{
    private ProjectNameSDK? _client;
    private Dictionary<string, object?>? _options;
    private int _netcalls;

    public TestFeature()
    {
        Version = "0.0.1";
        Name = "test";
        Active = true;
    }

    public override void Init(Context ctx, Dictionary<string, object?> options)
    {
        _client = ctx.Client;
        _options = options;

        var entity = Helpers.ToMapAny(StructUtils.GetProp(options, "entity"))
            ?? new Dictionary<string, object?>();

        _client!.Mode = "test";

        // Ensure entity ids are correct.
        StructUtils.Walk(entity, (key, val, parent, path) =>
        {
            if (path.Count == 2 && val is Dictionary<string, object?> m && key != null)
            {
                m["id"] = StructUtils.StrKey(key);
            }
            return val;
        });

        FetcherFunc testFetcher = (ctx2, _fullurl, _fetchdef) =>
        {
            static Dictionary<string, object?> Respond(int status, object? data,
                Dictionary<string, object?>? extra)
            {
                var res = new Dictionary<string, object?>
                {
                    ["status"] = status,
                    ["statusText"] = "OK",
                    ["json"] = (Func<object?>)(() => data),
                    ["body"] = "not-used",
                };
                if (extra != null)
                {
                    foreach (var kv in extra)
                    {
                        res[kv.Key] = kv.Value;
                    }
                }
                return res;
            }

            var op = ctx2.Op!;
            var entmap = Helpers.ToMapAny(StructUtils.GetProp(entity, op.Entity))
                ?? new Dictionary<string, object?>();

            // For single-entity ops (load, remove) with an empty explicit
            // match, fall back to the id the entity client already knows from
            // a prior create/load (in ctx.Match / ctx.Data). Mirrors the TS
            // mock where param() resolves the id from that accumulated state.
            Dictionary<string, object?> ResolveMatch(Dictionary<string, object?> explicitMatch)
            {
                if (explicitMatch.Count > 0)
                {
                    return explicitMatch;
                }
                foreach (var src in new object?[] { ctx2.Match, ctx2.Data })
                {
                    if (src == null)
                    {
                        continue;
                    }
                    var v = StructUtils.GetProp(src, "id");
                    if (v != null && !Equals(v, "__UNDEFINED__"))
                    {
                        return new Dictionary<string, object?> { ["id"] = v };
                    }
                }
                return new Dictionary<string, object?>();
            }

            if (op.Name == "load")
            {
                var args = BuildArgs(ctx2, op, ResolveMatch(ctx2.Reqmatch));
                var found = StructUtils.Select(entmap, args);
                var ent = StructUtils.GetElem(found, 0);
                if (ent == null)
                {
                    return Respond(404, null, new Dictionary<string, object?>
                    {
                        ["statusText"] = "Not found",
                    });
                }
                StructUtils.DelProp(ent, "$KEY");
                var outval = StructUtils.Clone(ent);
                return Respond(200, outval, null);
            }
            else if (op.Name == "list")
            {
                var args = BuildArgs(ctx2, op, ctx2.Reqmatch);
                var found = StructUtils.Select(entmap, args);
                if (found == null)
                {
                    return Respond(404, null, new Dictionary<string, object?>
                    {
                        ["statusText"] = "Not found",
                    });
                }
                foreach (var item in found)
                {
                    StructUtils.DelProp(item, "$KEY");
                }
                var outval = StructUtils.Clone(found);
                return Respond(200, outval, null);
            }
            else if (op.Name == "update")
            {
                // Match the existing entity by id only (or its alias).
                // Reqdata also contains the new field values, which would
                // otherwise cause Select to filter out the entity we want to
                // update. When reqdata has no id, fall back to the id the
                // entity client carries from a prior create/load.
                var updateMatch = new Dictionary<string, object?>();
                if (ctx2.Reqdata != null)
                {
                    if (ctx2.Reqdata.TryGetValue("id", out var idv))
                    {
                        updateMatch["id"] = idv;
                    }
                    if (op.Alias != null &&
                        StructUtils.GetProp(op.Alias, "id") is string aliasId &&
                        ctx2.Reqdata.TryGetValue(aliasId, out var av))
                    {
                        updateMatch[aliasId] = av;
                    }
                }
                if (updateMatch.Count == 0)
                {
                    updateMatch = ResolveMatch(new Dictionary<string, object?>());
                }
                var args = BuildArgs(ctx2, op, updateMatch);
                var found = StructUtils.Select(entmap, args);
                var ent = StructUtils.GetElem(found, 0);
                if (ent == null && entmap != null)
                {
                    foreach (var e in entmap.Values)
                    {
                        if (e is Dictionary<string, object?>)
                        {
                            ent = e;
                            break;
                        }
                    }
                }
                if (ent == null)
                {
                    return Respond(404, null, new Dictionary<string, object?>
                    {
                        ["statusText"] = "Not found",
                    });
                }
                if (ent is Dictionary<string, object?> entm && ctx2.Reqdata != null)
                {
                    foreach (var kv in ctx2.Reqdata)
                    {
                        entm[kv.Key] = kv.Value;
                    }
                }
                StructUtils.DelProp(ent, "$KEY");
                var outval = StructUtils.Clone(ent);
                return Respond(200, outval, null);
            }
            else if (op.Name == "remove")
            {
                var args = BuildArgs(ctx2, op, ResolveMatch(ctx2.Reqmatch));
                var found = StructUtils.Select(entmap, args);
                var ent = StructUtils.GetElem(found, 0);
                // Remove only the first matched entity. If nothing matches,
                // succeed as a no-op rather than erroring.
                if (ent is Dictionary<string, object?> entm2)
                {
                    var id = StructUtils.GetProp(entm2, "id");
                    StructUtils.DelProp(entmap, id);
                }
                return Respond(200, null, null);
            }
            else if (op.Name == "create")
            {
                _ = BuildArgs(ctx2, op, ctx2.Reqdata);
                var id = ctx2.Utility!.Param(ctx2, "id");
                id ??= string.Format("{0:x4}{1:x4}{2:x4}{3:x4}",
                    Random.Shared.Next(0x10000), Random.Shared.Next(0x10000),
                    Random.Shared.Next(0x10000), Random.Shared.Next(0x10000));

                var ent = StructUtils.Clone(ctx2.Reqdata);
                if (ent is Dictionary<string, object?> entm)
                {
                    entm["id"] = id;
                    if (id is string idStr)
                    {
                        entmap![idStr] = entm;
                    }
                    StructUtils.DelProp(entm, "$KEY");
                    var outval = StructUtils.Clone(entm);
                    return Respond(200, outval, null);
                }
                return Respond(200, ent, null);
            }

            return Respond(404, null, new Dictionary<string, object?>
            {
                ["statusText"] = "Unknown operation",
            });
        };

        // Optional network behaviour simulation over the mock transport.
        // Enable per test via `TestSDK(new(){["net"]=...}, null)`. When `net`
        // is absent the mock behaves exactly as before (no wrapping), so
        // existing generated tests are unaffected.
        var net = Helpers.ToMapAny(StructUtils.GetProp(options, "net"));
        ctx.Utility!.Fetcher = net == null ? testFetcher : MakeNetsim(net, testFetcher);
    }

    // MakeNetsim wraps a transport with simulated network conditions: latency
    // (fixed or {min,max}), a budget of first-N failures (`failTimes` ->
    // `failStatus`), first-N connection errors (`errorTimes`), or a hard
    // `offline` outage. Counter-driven, so simulations are deterministic
    // across a test.
    private FetcherFunc MakeNetsim(Dictionary<string, object?> net, FetcherFunc inner)
    {
        _netcalls = 0;

        int PickLatency()
        {
            if (!net.TryGetValue("latency", out var l) || l == null)
            {
                return 0;
            }
            if (l is Dictionary<string, object?> lm)
            {
                var min = FoptInt(lm, "min", 0);
                var max = FoptInt(lm, "max", min);
                if (max <= min)
                {
                    return min;
                }
                return min + ((max - min) >> 1);
            }
            var fixedMs = FoptInt(net, "latency", 0);
            return fixedMs < 0 ? 0 : fixedMs;
        }

        void SleepMs(int ms)
        {
            if (ms <= 0)
            {
                return;
            }
            FoptSleep(net)(ms);
        }

        return (ctx, url, fetchdef) =>
        {
            _netcalls++;
            var call = _netcalls;

            if (net.TryGetValue("offline", out var off) && off is bool netOffline && netOffline)
            {
                SleepMs(PickLatency());
                throw ctx.MakeError("netsim_offline",
                    "Simulated network offline (URL was: \"" + url + "\")");
            }
            if (call <= FoptInt(net, "errorTimes", 0))
            {
                SleepMs(PickLatency());
                throw ctx.MakeError("netsim_conn",
                    $"Simulated connection error (call {call})");
            }
            if (call <= FoptInt(net, "failTimes", 0))
            {
                SleepMs(PickLatency());
                var status = FoptInt(net, "failStatus", 503);
                return new Dictionary<string, object?>
                {
                    ["status"] = status,
                    ["statusText"] = "Simulated Failure",
                    ["body"] = "not-used",
                    ["json"] = (Func<object?>)(() => null),
                    ["headers"] = new Dictionary<string, object?>(),
                };
            }
            SleepMs(PickLatency());
            return inner(ctx, url, fetchdef);
        };
    }

    private static object BuildArgs(Context ctx, Operation op, Dictionary<string, object?>? args)
    {
        var opname = op.Name;

        // Get last point from config.
        var points = StructUtils.GetPath(ctx.Config,
            StructUtils.Jt("entity", ctx.Entity!.GetName(), "op", opname, "points"));
        var point = StructUtils.GetElem(points, -1);

        // Get required params.
        var paramsPath = StructUtils.GetPath(point, StructUtils.Jt("args", "params"));
        var reqdParams = StructUtils.Select(paramsPath,
            new Dictionary<string, object?> { ["reqd"] = true });
        var reqd = StructUtils.Transform(reqdParams,
            StructUtils.Jt("`$EACH`", "", "`$KEY.name`"));

        var qand = new List<object?>();
        var q = new Dictionary<string, object?> { ["`$AND`"] = qand };

        if (args != null)
        {
            foreach (var key in StructUtils.KeysOf(args))
            {
                var isId = key == "id";
                var selected = StructUtils.Select(reqd, key);
                var isReqd = !StructUtils.IsEmpty(selected);

                if (isId || isReqd)
                {
                    var v = ctx.Utility!.Param(ctx, key);
                    var ka = StructUtils.GetProp(op.Alias, key);

                    var qor = new List<object?>
                    {
                        new Dictionary<string, object?> { [key] = v },
                    };
                    if (ka is string kas)
                    {
                        qor.Add(new Dictionary<string, object?> { [kas] = v });
                    }

                    qand.Add(new Dictionary<string, object?> { ["`$OR`"] = qor });
                }
            }
        }

        if (ctx.Ctrl?.Explain != null)
        {
            ctx.Ctrl.Explain["test"] = new Dictionary<string, object?>
            {
                ["query"] = q,
            };
        }

        return q;
    }
}
